pkg/rhcos/release/release.go

```diff
@@ -0,0 +1,180 @@
+	imagev1 "github.com/openshift/api/image/v1"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+// RHCOSBuild extracts the RHCOS build version from an OpenShift
+// release image.
+func RHCOSBuild(ctx context.Context, releaseImagePullSpec string, pullSecret []byte) (string, error) {
+	imageReferencesBytes, err := imageReferences(ctx, releaseImagePullSpec, pullSecret)
+	if err != nil {
+		return "", err
+	}
+
+	pullSpec, err := osContentPullSpec(imageReferencesBytes)
+	if err != nil {
+		return "", err
+	}
+
+	imageSource, image, err := pull(ctx, pullSpec, pullSecret)
+	if err != nil {
+		return "", err
+	}
+	defer imageSource.Close()
+
+	inspect, err := image.Inspect(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	version, ok := inspect.Labels["version"]
+	if !ok {
+		return "", errors.Errorf("%s has no version annotation", pullSpec)
+	}
+
+	return version, nil
+}
+
+// pull pulls an image.  If the pull secret includes credentials for
+// the pull-spec authority, those are used to authenticate requests.
+// The signature on the returned image, if any, is validated agaist
+// the signature policy from /etc/containers/policy.json.
+func pull(ctx context.Context, pullSpec string, pullSecret []byte) (types.ImageSource, types.Image, error) {
+	named, err := reference.ParseNamed(pullSpec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	reference, err := docker.NewReference(named)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sys := &types.SystemContext{}
+	err = addPullSecret(sys, pullSecret, named)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	imageSource, err := reference.NewImageSource(ctx, sys)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer func() {
+		if err != nil {
+			imageSource.Close()
+		}
+	}()
+
+	unparsed := image.UnparsedInstance(imageSource, nil)
+
+	policy, err := signature.DefaultPolicy(sys)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if allowed, err := policyContext.IsRunningImageAllowed(ctx, unparsed); !allowed || err != nil {
+		return nil, nil, errors.Wrapf(err, "%s rejected", pullSpec)
+	}
+
+	img, err := image.FromUnparsedImage(ctx, sys, unparsed)
+	return imageSource, img, err
+}
+
+func imageReferences(ctx context.Context, releaseImagePullSpec string, pullSecret []byte) ([]byte, error) {
+	imageSource, image, err := pull(ctx, releaseImagePullSpec, pullSecret)
+	if err != nil {
+		return nil, err
+	}
+	defer imageSource.Close()
+
+	layers := image.LayerInfos()
+	for i := len(layers) - 1; i >= 0; i-- {
+		layer := layers[i]
+		blob, _, err := imageSource.GetBlob(ctx, layer, blobinfocache.NoCache)
+		if err != nil {
+			return nil, err
+		}
+		defer blob.Close()
+
+		// fmt.Println(layer.MediaType) // TODO: blank from quay?  Just assume this is a gzipped tar?
+		gzipReader, err := gzip.NewReader(blob)
```

Comment:
Consider `c/image/pkg/compression.DetectCompression`. Yes, there is a MIME type in the manifest, but historically it’s not been used to differentiate between compressed/uncompressed in practice.

Reply:
> Yes, there is a MIME type in the manifest...

Fix the broken servers! ;)  In practice, these pulls are going to be from release images that we cut with `oc adm release new ...` and the layer is going to be gzipped, but sure, I can reroll to use peek-inside detection.

Reply:
Sadly, it’s not just the servers; the schema2 spec does not officially specify an “uncompressed tar” MIME type, that’s new in OCI  (although there is a de-facto schema2 definition in containerd).

---

Actually, `…/compression.AutoDecompress` would probably be even more convenient.

I agree that in practice pretty much every image is compressed, and this may not be worth worrying about. Only noting that this exists because I saw that TODO.

Reply:
FWIW there is work ongoing to add new compression types like Zstd, so hard-coding `gzip` here will eventually break.